lib/common/output_xml.c

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright 2019 Chris Lumens <clumens@redhat.com>
+ *
+ * This source code is licensed under the GNU General Public License version 2
+ * or later (GPLv2+) WITHOUT ANY WARRANTY.
+ */
+
+#include <ctype.h>
+#include <stdlib.h>
+#include <crm/crm.h>
+#include <crm/common/output.h>
+#include <crm/common/output_internal.h>
+#include <crm/common/xml.h>
+
+typedef struct xml_private_s {
+    xmlNode *root;
+    GQueue *parent_q;
+} xml_private_t;
+
+static char *
+mangle_string(const char *in) {
+    char *out = strdup(in);
+
+    for (char *c = out; *c != '\0'; c++) {
+        if (*c == ' ')  {
+            *c = '-';
+        } else {
+            *c = tolower(*c);
+        }
+    }
```

Comment:
> (1) This function could potentially be exposed as internal API too.

This jogged my memory, we have some functions that aren't for this but could be reference points, see crm_xml_escape() and crm_xml_sanitize_id()

> (2) The naming rules don't like dashes. I could use underscores instead. We seem to use both, looking at the CIB.

dash = hyphen, it's acceptable just not as first character

> (3) I could put the name in a desc= attribute, yeah. I tend to like using elements instead of attributes but I don't have any strong reasons for doing so. We also use attributes extensively in the CIB, so maybe I should go that way to blend in.

Your call, you have to do the work :)

Reply:
> > (2) The naming rules don't like dashes. I could use underscores instead. We seem to use both, looking at the CIB.
> 
> dash = hyphen, it's acceptable just not as first character

Yeah I was referring to the best practices section a little farther down, which suggests not using dashes because some software may not like it.  But like I said, we use both in the CIB so it doesn't really matter.  I can do whichever looks better.

Reply:
Huh, I missed that. Their rationale sounds questionable, but let's avoid them to be on the safe side.